ISPN-11729 Expose rebalancing enabled values in the REST API

karesti opened the pull request:
https://issues.redhat.com/browse/ISPN-11729

ryanemerson commented on server/rest/src/main/java/org/infinispan/rest/resources/ContainerResource.java:
```diff
@@ -161,7 +161,18 @@ public Invocations getInvocations() {
+      try {
+      rebalancingEnabled = SecurityActions.getGlobalComponentRegistry(cacheManager).getLocalTopologyManager()
+            .isRebalancingEnabled();
+      } catch (Exception e) {
```
`SecurityException`?

Intellij is also warning about `e` being unused, so we should either add a comment in the body or change `e` to `ignore` to suppress the warning.

ryanemerson commented on server/rest/src/main/java/org/infinispan/rest/resources/ContainerResource.java:
```diff
@@ -161,7 +161,18 @@ public Invocations getInvocations() {
+      if (rebalancingEnabled != null) {
+         cacheManagerInfo.set("rebalancing_enabled", rebalancingEnabled);
+      }
+      return asJsonResponseFuture(cacheManagerInfo, responseBuilder);
```
I think we should add the rebalancing status to  `CacheManagerInfo` directly. That way embedded users can access this information as well and the `ContainerResource` is simpler.

karesti replied:
changed to cacheManagerinfo

ryanemerson reviewed (commented):
Couple of minor things.

ryanemerson commented on core/src/main/java/org/infinispan/manager/CacheManagerInfo.java:
```diff
@@ -127,6 +132,15 @@ public boolean isSiteCoordinator() {
+      }
+      return false;
+   }
+
```
I don't think this is right. If the rebalancing status cannot be retrieved it should not be set to `false` as this may not be the case, instead it should be null. That way the console (or other clients) can indicate that the status is actually unknown if `null` is returned.

So if you change this to be like your original implementation in the `ContainerResource`, where we return `Boolean` with null if a `Exception` occurs.

karesti replied:
According to the code, I really don't know when this is supposed to raise an exception and become a not redeable value

karesti commented:
@ryanemerson I added some docs and I also changed the caches list api that I use in the console because I realised I also need the value there to display the rebalancing enable/disable in the caches list as we planned for this feature

ryanemerson commented on core/src/main/java/org/infinispan/manager/CacheManagerInfo.java:
```diff
@@ -157,6 +170,12 @@ public Json toJson() {
             .set("sites_view", Json.make(getSites()));
+
+     if(isRebalancingEnabled() != null) {
+        result.set("rebalancing_enabled", isRebalancingEnabled());
```
I'm not sure the if is necessary. If the value is null, then `Json` will use the json nil object in it's place.

That being said, I see that we omit the key elsewhere ... so maybe better to be consistent.

karesti replied:
yes, sometimes keys are just not present since it might be a matter of being able to read some properties but not all of them (admin, observer...)
i rather omit and treat the value if it's present. if things are well documented I think it's fine. In any case, the principal consumer of these APIs are our own tools like CLI and Console ...

ryanemerson commented on core/src/main/java/org/infinispan/manager/CacheManagerInfo.java:
```diff
@@ -127,6 +131,15 @@ public boolean isSiteCoordinator() {
+      try {
+         return localTopologyManager.isRebalancingEnabled();
+      } catch (Exception e) {
+         // Do nothing
```
We can just return null in the catch.

ryanemerson commented on server/rest/src/main/java/org/infinispan/rest/resources/ContainerResource.java:
```diff
@@ -257,7 +260,26 @@ public Invocations getInvocations() {
+         } catch (Exception ex) {
+            // Something went wrong. Just ignore.
+         }
+      }
```
We can just make this a single try/catch:

```java
      LocalTopologyManager localTopologyManager = null;
      Boolean clusterRebalancingEnabled = null;
      try {
         localTopologyManager = SecurityActions.getGlobalComponentRegistry(cacheManager).getLocalTopologyManager();
         clusterRebalancingEnabled = localTopologyManager.isRebalancingEnabled();
      } catch (Exception e) {
         // Unable to get the component. Just ignore.
      }
```

karesti replied:
ok

ryanemerson commented:
Checkstyle is failing, other than that just a couple of nitpicks.

gustavonalle commented on documentation/src/main/asciidoc/topics/ref_rest_cache_managers.adoc:
```diff
@@ -36,6 +36,7 @@ If cross-site replication is not configured, {brandname} returns "local".
 * `site_coordinators_address` is an array of logical addresses of site coordinators.
 * `sites_view` The list of sites that participate in cross-site replication. +
 If cross-site replication is not configured, {brandname} returns an empty list.
+* `rebalancing_enabled` is true if rebalancing is enabled. Fetching this property might fail on the server. In that case the property won't be present in the payload.
```
Out of curiosity, why it might fail? This doesn't inspire a lot of confidence :)

karesti replied:
the method raises exception that can raise by a timeout for example because executes a command

danberindei replied:
Servers other than the cluster coordinator have to connect to the coordinator to fetch the status.
It does retry if the coordinator crashes, so I think the only possibility is that the coordinator doesn't reply and the RPC times out.

ryanemerson commented:
@karesti Something has gone very wrong ... `867 failures` :scream:

karesti commented:
@ryanemerson yes, I think it's related to putting the value in the json

karesti commented:
@ryanemerson trying a fix

danberindei commented on server/rest/src/main/java/org/infinispan/rest/resources/CacheResourceV2.java:
```diff
@@ -548,6 +550,16 @@ private RestResponse getDetailResponse(Cache<?, ?> cache) {
 
+      Boolean rebalancingEnabled = null;
+      try {
+         if(cache instanceof AbstractDelegatingAdvancedCache) {
```
I have a hunch this condition is not true for secured caches. It would be better if you used SecurityActions to extract the `LocalTopologyManager` via the component registry and invoke `isRebalancingEnabled()` on it.

karesti replied:
changing that

danberindei commented on server/rest/src/main/java/org/infinispan/rest/resources/ContainerResource.java:
```diff
@@ -257,7 +259,18 @@ public Invocations getInvocations() {
+      }
+
       // We rely on the fact that getCacheNames doesn't block for embedded - remote it does unfortunately
+      LocalTopologyManager finalLocalTopologyManager = localTopologyManager;
```
You wouldn't need new variables if you initialized `localTopologyManager` and `clusterRebalancingEnabled` inside the try + catch blocks.

karesti commented:
I pushed all the latests changes @ryanemerson @danberindei

ryanemerson reviewed (approved):
Just waiting on CI :crossed_fingers:

karesti commented:
@ryanemerson I released and did the console pr as well

danberindei commented:
Merged, thanks Katia!
